@taskgroup decorator implementation  [https://github.com/apache/airflow/issues/11870 ]

VBhojawala opened the pull request:
New Feature of @taskgroup decorator implementation 
closes:#11870

Added a decorator function which can be used to create TaskGroup for python function.

I have tested it with following example : 

```py
def generate_value():
    """Dummy function"""
    return "Bring me a shrubbery!"


@task()
def print_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Print Value: %s (at %s)", value, ctx['ts'])
    return value


@task()
def log_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Log Value: %s (at %s)", value, ctx['ts'])


@taskgroup(group_id='t3_task_1')
def t3_task_1(value):
    op1 = print_value(value)
    log_value(op1)


with DAG(
    dag_id='T3_task_group_deco',
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
    tags=['example'],
) as dag:
    task1 = PythonOperator(
        task_id='generate_value',
        python_callable=generate_value,
    )

    t3_task_1(task1.output)
```

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

VBhojawala committed:
Update task_group.py

New Feature of @taskgroup decorator implementation 
https://github.com/apache/airflow/issues/11870 

Added a decorator function which can be used to create TaskGroup.

I have tested it with following example : 

def generate_value():
    """Dummy function"""
    return "Bring me a shrubbery!"


@task()
def print_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Print Value: %s (at %s)", value, ctx['ts'])
    return value


@task()
def log_value(value):
    """Dummy function"""
    ctx = get_current_context()
    log.info("Log Value: %s (at %s)", value, ctx['ts'])


@taskgroup(group_id='t3_task_1')
def t3_task_1(value):
    op1 = print_value(value)
    log_value(op1)


with DAG(
    dag_id='T3_task_group_deco',
    default_args={'owner': 'airflow'},
    start_date=days_ago(2),
    schedule_interval=None,
    tags=['example'],
) as dag:
    task1 = PythonOperator(
        task_id='generate_value',
        python_callable=generate_value,
    )

    t3_task_1(task1.output)

turbaszek reviewed (changes requested):
Generally looks good! However, we will need some test and documentation + examples around that.

VBhojawala committed:
Added @functools.wraps(f)

VBhojawala committed:
Example of @taskgroup decorator

Added an Example demonstrating use of @taskgroup decorator

VBhojawala committed:
Added @taskgroup decorator documentation

VBhojawala committed:
Update example_task_group_decorator.py

turbaszek commented on docs/concepts.rst:
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
 
 .. image:: img/task_group.gif
 
+@taskgroup decorator
```
```suggestion
TaskGroup decorator
```

turbaszek commented on docs/concepts.rst:
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
+@taskgroup decorator
+--------------------
+
+TaskGroup can be created using @taskgroup decorator, it takes one mandatory argument 'group_id' which is same as constructor of TaskGroup class. It works exactly same as creating TaskGroup using context manager 'with TaskGroup('groupid') as section:'.
```
```suggestion
TaskGroup can be created using ``@taskgroup decorator``, it takes one mandatory argument ``group_id`` which is same as constructor of TaskGroup class. It works exactly same as creating TaskGroup using context manager ``with TaskGroup('groupid') as section:``.
```

turbaszek commented on docs/concepts.rst:
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
+
+  @task
+  def task_2(value):
+      print( f'[ Task2 {value} ]')
```
```suggestion
      print(f'[ Task2 {value} ]')
```

turbaszek commented on docs/concepts.rst:
```diff
@@ -1017,6 +1017,32 @@ This animated gif shows the UI interactions. TaskGroups are expanded or collapse
+  @taskgroup(group_id='section_1')
+  def section_1(value):
+      return task_2(task_1(value))
+
```
I'm not sure if we need two examples. If we decide to keep both the it would be nice to add some text between them.

turbaszek reviewed (changes requested):
@VBhojawala we still needs tests for this decorator to avoid regression in future

VBhojawala committed:
Update docs/concepts.rst

Co-authored-by: Tomek Urbaszek <turbaszek@gmail.com>

VBhojawala committed:
Update docs/concepts.rst

Co-authored-by: Tomek Urbaszek <turbaszek@gmail.com>

VBhojawala committed:
Removed second taskgroup from example

VBhojawala committed:
Added Test cases for TaskGroup decorator

VBhojawala committed:
Grou_id copy from function name

Added code to set group_id of TaskGroup from function name in case of no positional argument or keyword argument with key 'group_id' is provided.

VBhojawala committed:
Updated test cases 

Updated test case fro taskgroup decorator without group_id

VBhojawala committed:
Added group_id is copied from function name

VBhojawala committed:
Added documentation pointers

VBhojawala committed:
added example pointers

VBhojawala committed:
Updated for static checks

VBhojawala committed:
updated for static checks

VBhojawala committed:
added support for only @taskgroup decorator 

added support for only @taskgroup decorator instead of  @taskgroup()

VBhojawala committed:
added taskgroup for @taskgroup decorator
